### ISSUE TITLE
Elasticsearch index health turned Red!!!

### ISSUE THREAD

**Issue by abhi-paul** [elastic/elasticsearch#24358]
Hi All,
My Environment details,
Used ELK Version 5.1.2, with single node as master in the elasticsearch cluster.
**After index got created in elasticsearch few hours after health of that index turned Red !!!**
During my Observation few suspected item are:
```
>Shard Allocation Failed
>org.apache.lucene.index.**CorruptIndexException: codec footer mismatch** (file truncated?): actual footer=0 vs expected footer=-1        071082520 (resource=BufferedChecksumIndexInput(MMapIndexInput(path="/var/elasticsearch/nodes/0/indices/L2Z0jkfDQkyHc0q2MDfVAA        /2/index/_49f_Lucene50_0.tim")))
>hostname hostip - WARN - elasticsearch[master][refresh][T#1] - - - [org.elasticsearch.index.IndexService] [master] [logstash-        2017.04.26] `failed to run task refresh - suppressing re-occurring exceptions unless the exception changes`
  org.elasticsearch.index.engine.RefreshFailedEngineException: Refresh failed
          at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:658) ~[elasticsearch-5.1.2.jar:5.1.2]
```

My Observation as follows,
Observation 1:
```
> GET _cat/indices?v
health status index               uuid                   pri rep docs.count docs.deleted store.size pri.store.size
red    open   logstash-2017.04.27 tdl7RrQRT36-azv9L5bMMw   5   1      51146            0     39.6mb         39.6mb
red    open   logstash-2017.04.25 L2Z0jkfDQkyHc0q2MDfVAA   5   1     196206            0     83.1mb         83.1mb
yellow open   .kibana             Y7x_PvHVTKuHTUVPw1YrQg   5   1          5            0     50.1kb         50.1kb
red    open   logstash-2017.04.26 LnJhX2ZDSHeqcHvk99OoVQ   5   1      81793            0     33.8mb         33.8mb
```
Observation 2:
```
>POST /_flush/synced
{
  "_shards": {
    "total": 40,
    "successful": 15,
    "failed": 10
  },
  ".kibana": {
    "total": 10,
    "successful": 5,
    "failed": 0
  },
  "logstash-2017.04.25": {
    "total": 10,
    "successful": 3,
    "failed": 4,
    "failures": [
      {
        "shard": 2,
        "reason": "no active shards"
      },
      {
        "shard": 3,
        "reason": "no active shards"
      }
    ]
  },
  "logstash-2017.04.26": {
    "total": 10,
    "successful": 3,
    "failed": 4,
    "failures": [
      {
        "shard": 0,
        "reason": "no active shards"
      },
      {
        "shard": 4,
        "reason": "no active shards"
      }
    ]
  },
  "logstash-2017.04.27": {
    "total": 10,
    "successful": 4,
    "failed": 2,
    "failures": [
      {
        "shard": 0,
        "reason": "no active shards"
      }
    ]
  }
}
```
Observation 3:
```
curl -XGET 172.31.148.158:9200/_cat/shards?h=index,shard,prirep,state,unassigned.reason| grep UNASSIGNED
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1785  100  1785    0     0  11711      0 --:--:-- --:--:-- --:--:-- 11821
.kibana             4 r UNASSIGNED INDEX_CREATED
.kibana             1 r UNASSIGNED INDEX_CREATED
.kibana             3 r UNASSIGNED INDEX_CREATED
.kibana             2 r UNASSIGNED INDEX_CREATED
.kibana             0 r UNASSIGNED INDEX_CREATED
logstash-2017.04.27 4 r UNASSIGNED INDEX_CREATED
logstash-2017.04.27 1 r UNASSIGNED INDEX_CREATED
logstash-2017.04.27 3 r UNASSIGNED INDEX_CREATED
logstash-2017.04.27 2 r UNASSIGNED INDEX_CREATED
logstash-2017.04.27 0 p UNASSIGNED ALLOCATION_FAILED
logstash-2017.04.27 0 r UNASSIGNED INDEX_CREATED
logstash-2017.04.26 4 p UNASSIGNED ALLOCATION_FAILED
logstash-2017.04.26 4 r UNASSIGNED INDEX_CREATED
logstash-2017.04.26 1 r UNASSIGNED INDEX_CREATED
logstash-2017.04.26 3 r UNASSIGNED INDEX_CREATED
logstash-2017.04.26 2 r UNASSIGNED INDEX_CREATED
logstash-2017.04.26 0 p UNASSIGNED ALLOCATION_FAILED
logstash-2017.04.26 0 r UNASSIGNED INDEX_CREATED
logstash-2017.04.25 4 r UNASSIGNED INDEX_CREATED
logstash-2017.04.25 1 r UNASSIGNED INDEX_CREATED
logstash-2017.04.25 3 p UNASSIGNED ALLOCATION_FAILED
logstash-2017.04.25 3 r UNASSIGNED INDEX_CREATED
logstash-2017.04.25 2 p UNASSIGNED ALLOCATION_FAILED
logstash-2017.04.25 2 r UNASSIGNED INDEX_CREATED
logstash-2017.04.25 0 r UNASSIGNED INDEX_CREATED
```
Observation 4:
```
"logstash-2017.04.25": {
      "status": "red",
      "number_of_shards": 5,
      "number_of_replicas": 1,
      "active_primary_shards": 3,
      "active_shards": 3,
      "relocating_shards": 0,
      "initializing_shards": 0,
      "unassigned_shards": 7,
      "shards": {
        "0": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "1": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "2": {
          "status": "red",
          "primary_active": false,
          "active_shards": 0,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 2
        },
        "3": {
          "status": "red",
          "primary_active": false,
          "active_shards": 0,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 2
        },
        "4": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        }
      }
    },
    "logstash-2017.04.26": {
      "status": "red",
      "number_of_shards": 5,
      "number_of_replicas": 1,
      "active_primary_shards": 3,
      "active_shards": 3,
      "relocating_shards": 0,
      "initializing_shards": 0,
      "unassigned_shards": 7,
      "shards": {
        "0": {
          "status": "red",
          "primary_active": false,
          "active_shards": 0,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 2
        },
        "1": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "2": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "3": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "4": {
          "status": "red",
          "primary_active": false,
          "active_shards": 0,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 2
        }
      }
    },
    "logstash-2017.04.27": {
      "status": "red",
      "number_of_shards": 5,
      "number_of_replicas": 1,
      "active_primary_shards": 4,
      "active_shards": 4,
      "relocating_shards": 0,
      "initializing_shards": 0,
      "unassigned_shards": 6,
      "shards": {
        "0": {
          "status": "red",
          "primary_active": false,
          "active_shards": 0,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 2
        },
        "1": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "2": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "3": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        },
        "4": {
          "status": "yellow",
          "primary_active": true,
          "active_shards": 1,
          "relocating_shards": 0,
          "initializing_shards": 0,
          "unassigned_shards": 1
        }
      }
```
Observation 5:
Tried force a shard to be assigned
```
[root@sys]# curl -XPOST '172.31.148.158:9200/_cluster/reroute' -d '{ "commands" :
>   [ { "allocate" :
>       { "index" : "logstash-2017.04.25", "shard" : 3, "node": "master" }
>   }]
> }'
{"error":{"root_cause":[{"type":"parsing_exception","reason":"no [allocation_command] registered for [allocate]","line":3,"col":7}],"type":"parsing_exception","reason":"[cluster_reroute] failed to parse field [commands]","line":3,"col":7,"caused_by":{"type":"parsing_exception","reason":"no [allocation_command] registered for [allocate]","line":3,"col":7}},"status":400}
[root@sys]#
```
Observation 6:
Few suspected log entries found in elasticsearch log as below,
```
**LOG:** hostname hostip - WARN - elasticsearch[master][refresh][T#1] - - - [org.elasticsearch.index.engine.Engine] [master] [logstash-2017.04.25][2] failed engine [refresh failed]
org.apache.lucene.index.CorruptIndexException: codec footer mismatch (file truncated?): actual footer=0 vs expected footer=-1071082520 (resource=BufferedChecksumIndexInput(MMapIndexInput(path="/var/elasticsearch/nodes/0/indices/L2Z0jkfDQkyHc0q2MDfVAA/2/index/_49f_Lucene50_0.tim")))
        at org.apache.lucene.codecs.CodecUtil.validateFooter(CodecUtil.java:499) ~[lucene-core-6.3.0.jar:6.3.0 a66a44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
        at org.apache.lucene.codecs.CodecUtil.checkFooter(CodecUtil.java:411) ~[lucene-core-6.3.0.jar:6.3.0 a66a44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]

**LOG:**hostname hostip - WARN - elasticsearch[master][generic][T#5] - - - [org.elasticsearch.indices.cluster.IndicesClusterStateService] [master] [[logstash-2017.04.25][2]] marking and sending shard failed due to [shard failure, reason [refresh failed]]
 145813 org.apache.lucene.index.CorruptIndexException: codec footer mismatch (file truncated?): actual footer=0 vs expected footer=-1        071082520 (resource=BufferedChecksumIndexInput(MMapIndexInput(path="/var/elasticsearch/nodes/0/indices/L2Z0jkfDQkyHc0q2MDfVAA        /2/index/_49f_Lucene50_0.tim")))
          at org.apache.lucene.codecs.CodecUtil.validateFooter(CodecUtil.java:499) ~[lucene-core-6.3.0.jar:6.3.0 a66a44513ee819        1e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
          at org.apache.lucene.codecs.CodecUtil.checkFooter(CodecUtil.java:411) ~[lucene-core-6.3.0.jar:6.3.0 a66a44513ee8191e2        5b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]

**LOG:**hostname hostip - WARN - elasticsearch[master][refresh][T#1] - - - [org.elasticsearch.index.IndexService] [master] [logstash-        2017.04.26] failed to run task refresh - suppressing re-occurring exceptions unless the exception changes
  org.elasticsearch.index.engine.RefreshFailedEngineException: Refresh failed
          at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:658) ~[elasticsearch-5.1.2.jar:5.1.2]
          at org.elasticsearch.index.shard.IndexShard.refresh(IndexShard.java:629) ~[elasticsearch-5.1.2.jar:5.1.2]

**LOG:**hostname hostip - WARN - elasticsearch[master][generic][T#1] - - - [org.elasticsearch.cluster.action.shard.ShardStateAction]         [master] [logstash-2017.04.26][4] received shard failed for shard id [[logstash-2017.04.26][4]], allocation id [8xd6MYHSRM6DTFp4iVcLqg], primary term [0], message [shard failure, reason [refresh failed]], failure [CorruptIndexException[compound sub-files must have a valid codec header and footer: file is too small (0 bytes) (resource=BufferedChecksumIndexInput(MMapIndexInp        ut(path="/var/elasticsearch/nodes/0/indices/LnJhX2ZDSHeqcHvk99OoVQ/4/index/_5v_Lucene54_0.dvd")))]]
 346317 org.apache.lucene.index.CorruptIndexException: compound sub-files must have a valid codec header and footer: file is too smal        l (0 bytes) (resource=BufferedChecksumIndexInput(MMapIndexInput(path="/var/elasticsearch/nodes/0/indices/LnJhX2ZDSHeqcHvk99OoVQ/4/index/_5v_Lucene54_0.dvd")))
          at org.apache.lucene.codecs.CodecUtil.verifyAndCopyIndexHeader(CodecUtil.java:282) ~[lucene-core-6.3.0.jar:6.3.0 a66a        44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
          at org.apache.lucene.codecs.lucene50.Lucene50CompoundFormat.write(Lucene50CompoundFormat.java:96) ~[lucene-core-6.3.0        .jar:6.3.0 a66a44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]

**LOG:**hostname hostip - DEBUG - elasticsearch[master][bulk][T#1] - - - [org.elasticsearch.action.bulk.TransportShardBulkAction] [master] [logstash-2017.04.25][0] failed to execute bulk item (index) index {[logstash-2017.04.25][syslog][AVul0yU8Yq5pi7b0GTla], source[{"sourceapp":"kernel","offset":546316,"datacenterid":"DCA","type":"syslog","messagetimestamp":"Apr 25 15:55:25","targetip":"-","log_type":"kafka","messagebody":"XFS (dm-34): Ending clean mount","sourceappver":"-","@timestamp":"2017-04-25T15:55:25.000Z","instancename":"ost-abhi-ost3120-elkv4-nodebfyd","targethost":"ost-abhi-ost3120-elkv4-nodebfyd","beat":{"hostname":"filebeat-wkhjf","name":"filebeat-wkhjf","version":"5.1.2"},"@version":"1","instancetype":"dev","logfilename":"/var/log/messages"}]}
org.elasticsearch.index.mapper.MapperParsingException: failed to parse [messagetimestamp]
        at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:297) ~[elasticsearch-5.1.2.jar:5.1.2]
        at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:438) ~[elasticsearch-5.1.2.jar:5.1.2]

**LOG:**hostname hostip - WARN - elasticsearch[master][generic][T#1] - - - [org.elasticsearch.cluster.action.shard.ShardStateAction]         [master] [logstash-2017.04.26][4] received shard failed for shard id [[logstash-2017.04.26][4]], allocation id [8xd6MYHSRM6DTFp4iVcLqg], primary term [0], message [shard failure, reason [refresh failed]], failure [CorruptIndexException[compound sub-files must have a valid codec header and footer: file is too small (0 bytes) (resource=BufferedChecksumIndexInput(MMapIndexInp        ut(path="/var/elasticsearch/nodes/0/indices/LnJhX2ZDSHeqcHvk99OoVQ/4/index/_5v_Lucene54_0.dvd")))]]
 346317 org.apache.lucene.index.CorruptIndexException: compound sub-files must have a valid codec header and footer: file is too smal        l (0 bytes) (resource=BufferedChecksumIndexInput(MMapIndexInput(path="/var/elasticsearch/nodes/0/indices/LnJhX2ZDSHeqcHvk99OoVQ/4/index/_5v_Lucene54_0.dvd")))
          at org.apache.lucene.codecs.CodecUtil.verifyAndCopyIndexHeader(CodecUtil.java:282) ~[lucene-core-6.3.0.jar:6.3.0 a66a        44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
          at org.apache.lucene.codecs.lucene50.Lucene50CompoundFormat.write(Lucene50CompoundFormat.java:96) ~[lucene-core-6.3.0        .jar:6.3.0 a66a44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
```
Any ideas why it's happening? Is there a guide for recovering from this index health RED to Yellow/Green?

**Comment by danielmitterdorfer** [comment]
Hi @abhi-paul 

> Any ideas why it's happening?

Seems you upgraded these indices from Elasticsearch 2 (2.2.0?) and something went wrong during the upgrade? Some of your index file are corrupt as is indicated by messages like `codec footer mismatch (file truncated?): actual footer=0 vs expected footer=-1071082520`.

We also reserve Github for bug reports and feature requests only. Please ask questions like these in the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch) instead. Thank you!

**Comment by danielmitterdorfer** [comment]
For completeness and future reference, this issue happened on GlusterFS < 3.10.0 and there is [a related issue 1390050](https://bugzilla.redhat.com/show_bug.cgi?id=1390050) which has been fixed in GlusterFS 3.10.0 ([source](https://discuss.elastic.co/t/elasticsearch-index-health-turned-red/83805/5))
